iree/compiler/Dialect/LinalgExt/Transforms/Tiling.cpp

```diff
@@ -0,0 +1,415 @@
+  if (isUntiledLoop(tileSizes[loopDepth])) {
+    auto zeroAttr = builder.getI64IntegerAttr(0);
+    offsets.push_back(zeroAttr);
+    assert(matchPattern(loopBounds[loopDepth].offset, m_Zero()) &&
+           "expected loop bounds to have lower bound of zero");
+    tileSizes[loopDepth] = getOpFoldResult(loopBounds[loopDepth].size);
+    return tileLinalgExtOpImpl(builder, op, outputs, tileSizes, iteratorTypes,
+                               loopBounds, loopDepth + 1, offsets,
+                               distributionInfo);
+  }
+
+  // Generate an scf.for for the current loop depth.
+  Value lb = loopBounds[loopDepth].offset;
+  Value ub = loopBounds[loopDepth].size;
+  if (!matchPattern(loopBounds[loopDepth].stride, m_One())) {
+    return static_cast<LogicalResult>(
+        op.emitOpError("expected stride to be 1"));
+  }
+  Value step = getValue(builder, loc, tileSizes[loopDepth]);
+
+  // Update lb, ub and step for cyclic distribution.
+  if (!distributionInfo.empty() &&
+      iteratorTypes[loopDepth] == getParallelIteratorTypeName()) {
+    linalg::updateBoundsForCyclicDistribution(
+        builder, loc, distributionInfo.front().procId,
+        distributionInfo.front().nprocs, lb, ub, step);
+    distributionInfo = distributionInfo.drop_front();
+  }
+  FailureOr<TiledOp> innerReturnValue;
+  bool isBufferTiling = op->getNumResults() == 0;
+  ValueRange initValues(isBufferTiling ? ValueRange{} : outputs);
+  auto forOp = builder.create<scf::ForOp>(
+      loc, lb, ub, step, initValues,
+      [&](OpBuilder &b, Location loc, Value iv, ValueRange args) {
+        offsets.push_back(iv);
+        auto affineMaps = AffineMap::inferFromExprList({ArrayRef<AffineExpr>{
+            b.getAffineSymbolExpr(0),
+            b.getAffineSymbolExpr(1) - b.getAffineDimExpr(0)}})[0];
+        Value inBoundsTileSize = b.create<AffineMinOp>(
+            loc, affineMaps,
+            ValueRange{iv, getValue(builder, loc, tileSizes[loopDepth]), ub});
+        tileSizes[loopDepth] = getOpFoldResult(inBoundsTileSize);
+        // Recursively proceed to generate the tiled loop for the next level.
+        innerReturnValue = tileLinalgExtOpImpl(
+            b, op, (isBufferTiling ? outputs : args), tileSizes, iteratorTypes,
+            loopBounds, loopDepth + 1, offsets, distributionInfo);
+        if (failed(innerReturnValue)) return;
+        b.create<scf::YieldOp>(loc, innerReturnValue->results);
+      });
+  if (failed(innerReturnValue)) {
+    return innerReturnValue;
+  }
+  innerReturnValue->loops.insert(innerReturnValue->loops.begin(),
+                                 forOp.getOperation());
+  innerReturnValue->results = forOp.getResults();
+  return innerReturnValue;
+}
+
+FailureOr<TiledOp> tileLinalgExtOp(OpBuilder &b, LinalgExtOp op,
+                                   const linalg::LinalgTilingOptions &options) {
+  TiledOpInterface tilableOp = dyn_cast<TiledOpInterface>(op.getOperation());
+  if (!tilableOp) return TiledOp{};
```

Comment:
I can not get the behavior here. Is it equivalent to `return failure()`? If so, can we use `return failure()`? instead?

Reply:
`FailureOr<..>` is meant to capture an actual failure. If the op doesnt implement the `TiledOpInterface` it is not an error. The pattern should just return failure (there was a bug below, for handling this case. Fixed that). We need three states basically, Success, failure or do nothing. Returning a `TiledOp{}` is essentially do nothing case for the caller to resolve.

Reply:
Got you, thanks!